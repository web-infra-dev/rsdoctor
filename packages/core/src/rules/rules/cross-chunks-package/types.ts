export interface Config {
  ignore: string[];
}
