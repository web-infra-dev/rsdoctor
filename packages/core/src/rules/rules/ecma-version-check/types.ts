import type { PluginCheckSyntaxOptions } from '@rsbuild/plugin-check-syntax';

export interface Config extends PluginCheckSyntaxOptions {}
