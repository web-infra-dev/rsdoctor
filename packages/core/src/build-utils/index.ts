export * as Build from './build';
