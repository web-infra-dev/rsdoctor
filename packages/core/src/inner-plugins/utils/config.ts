import { Common, Linter, Plugin, SDK } from '@rsdoctor/types';
import assert from 'assert';
import type {
  RuleSetCondition as RspackRuleSetCondition,
  RuleSetRule as RspackRuleSetRule,
} from '@rspack/core';
import {
  RuleSetCondition as WebpackRuleSetCondition,
  RuleSetConditionAbsolute as WebpackRuleSetConditionAbsolute,
  RuleSetRule as WebpackRuleSetRule,
} from 'webpack';
import {
  RsdoctorWebpackPluginOptions,
  RsdoctorPluginOptionsNormalized,
  IReportCodeType,
  RsdoctorRspackPluginOptions,
  RsdoctorRspackPluginOptionsNormalized,
} from '@/types';
import { chalk, logger } from '@rsdoctor/utils/logger';

function defaultBoolean(v: unknown, dft: boolean): boolean {
  return typeof v === 'boolean' ? v : dft;
}

function getDefaultOutput() {
  return {
    reportCodeType: {
      noModuleSource: false,
      noAssetsAndModuleSource: false,
      noCode: false,
    },
    reportDir: '',
    compressData: true,
  };
}

function getDefaultSupports() {
  return {
    parseBundle: true,
    banner: undefined,
    gzip: true, // change the gzip to true by default.
  };
}

function normalizeFeatures(features: any, mode: keyof typeof SDK.IMode) {
  if (Array.isArray(features)) {
    return {
      loader: features.includes('loader'),
      plugins: features.includes('plugins'),
      resolver: features.includes('resolver'),
      bundle: features.includes('bundle'),
      treeShaking: features.includes('treeShaking'),
      lite: features.includes('lite') || mode === SDK.IMode[SDK.IMode.lite],
    };
  }
  return {
    loader: defaultBoolean(features.loader, true),
    plugins: defaultBoolean(features.plugins, true),
    resolver: defaultBoolean(features.resolver, false),
    bundle: defaultBoolean(features.bundle, true),
    treeShaking: defaultBoolean(features.treeShaking, false),
    lite:
      defaultBoolean(features.lite, false) ||
      mode === SDK.IMode[SDK.IMode.lite],
  };
}

function normalizeLinter(linter: any) {
  return {
    rules: {} as any,
    extends: [] as any,
    level: 'Error',
    ...linter,
  };
}

export function normalizeUserConfig<Rules extends Linter.ExtendRuleData[]>(
  config: RsdoctorWebpackPluginOptions<Rules> = {},
): RsdoctorPluginOptionsNormalized<Rules> {
  const {
    linter = {},
    features = {},
    loaderInterceptorOptions = {},
    disableClientServer = false,
    sdkInstance,
    innerClientPath = '',
    output = getDefaultOutput(),
    supports = getDefaultSupports(),
    port,
    printLog = { serverUrls: true },
    mode = 'normal',
    brief = {
      reportHtmlDir: undefined,
      reportHtmlName: undefined,
      writeDataJson: false,
    },
  } = config;

  assert(typeof linter === 'object');
  assert(typeof features === 'object' || Array.isArray(features));
  assert(typeof loaderInterceptorOptions === 'object');
  assert(typeof disableClientServer === 'boolean');

  const _features = normalizeFeatures(features, mode);
  const _linter = normalizeLinter(linter);

  // If lite mode is enabled and mode is not brief, set mode to lite
  let _mode: keyof typeof SDK.IMode = mode;

  if (_features.lite && _mode !== SDK.IMode[SDK.IMode.brief]) {
    _mode = SDK.IMode[SDK.IMode.lite] as keyof typeof SDK.IMode;
  }

  const reportCodeType = output.reportCodeType
    ? normalizeReportType(output.reportCodeType, _mode)
    : normalizeReportType(getDefaultOutput().reportCodeType, _mode);

  const res: RsdoctorPluginOptionsNormalized<Rules> = {
    linter: _linter,
    features: _features,
    loaderInterceptorOptions: {
      skipLoaders: Array.isArray(loaderInterceptorOptions.skipLoaders)
        ? loaderInterceptorOptions.skipLoaders
        : [],
    },
    disableClientServer,
    sdkInstance,
    output: {
      reportCodeType,
      reportDir: output.reportDir || '',
      compressData:
        output.compressData !== undefined ? output.compressData : true,
    },
    innerClientPath,
    supports,
    port,
    printLog,
    mode,
    brief,
  };

  if (
    res.output.compressData === false &&
    (!output.reportCodeType || !output.reportCodeType.noAssetsAndModuleSource)
  ) {
    logger.info(
      chalk.yellow(
        `[RSDOCTOR]: When you use compressData: false, it is recommended to set output.reportCodeType to { noAssetsAndModuleSource: true }.`,
      ),
    );
  }

  return res;
}

export function makeRuleSetSerializable(
  item:
    | RspackRuleSetCondition
    | WebpackRuleSetConditionAbsolute
    | WebpackRuleSetCondition
    | void,
) {
  if (!item) return;

  if (item instanceof RegExp) {
    // Used by the JSON.stringify method to enable the transformation of an object's data for JavaScript Object Notation (JSON) serialization.
    (item as Common.PlainObject).toJSON = item.toString;
  } else if (Array.isArray(item)) {
    item.forEach((i) => makeRuleSetSerializable(i));
  } else if (typeof item === 'object') {
    makeRuleSetSerializable(item.and);
    makeRuleSetSerializable(item.or);
    makeRuleSetSerializable(item.not);
  }
}

export function makeRulesSerializable(
  rules:
    | Plugin.RuleSetRule[]
    | RspackRuleSetRule['oneOf']
    | WebpackRuleSetRule['oneOf'],
) {
  if (!Array.isArray(rules)) return;

  if (!rules.length) return;

  rules.forEach((rule) => {
    if (!rule) return;

    makeRuleSetSerializable(rule.test);
    makeRuleSetSerializable(rule.resourceQuery);
    makeRuleSetSerializable(rule.resource);
    makeRuleSetSerializable(rule.resourceFragment);
    makeRuleSetSerializable(rule.scheme);
    makeRuleSetSerializable(rule.issuer);

    if ('issuerLayer' in rule) {
      makeRuleSetSerializable(rule.issuerLayer);
    }

    makeRuleSetSerializable(rule.include);
    makeRuleSetSerializable(rule.exclude);

    if (rule.oneOf) {
      makeRulesSerializable(rule.oneOf);
    }

    if ('rules' in rule && rule.rules) {
      makeRulesSerializable(rule.rules);
    }
  });
}

export const normalizeReportType = (
  reportCodeType: IReportCodeType,
  mode: keyof typeof SDK.IMode,
): SDK.ToDataType => {
  if (reportCodeType.noCode) {
    return SDK.ToDataType.NoCode;
  }

  if (mode === SDK.IMode[SDK.IMode.brief]) {
    return SDK.ToDataType.NoCode;
  }

  if (mode === SDK.IMode[SDK.IMode.lite]) {
    return SDK.ToDataType.NoSourceAndAssets;
  }

  if (reportCodeType.noAssetsAndModuleSource) {
    return SDK.ToDataType.NoSourceAndAssets;
  }

  if (reportCodeType.noModuleSource) {
    return SDK.ToDataType.NoSource;
  }

  return SDK.ToDataType.Normal;
};

export function normalizeRspackUserOptions<
  Rules extends Linter.ExtendRuleData[],
>(
  options: RsdoctorRspackPluginOptions<Rules>,
): RsdoctorRspackPluginOptionsNormalized<Rules> {
  const config: RsdoctorRspackPluginOptionsNormalized<Rules> =
    normalizeUserConfig(options);
  config.experiments ??= {};
  config.experiments.enableNativePlugin =
    options.experiments?.enableNativePlugin ?? false;
  return config;
}
