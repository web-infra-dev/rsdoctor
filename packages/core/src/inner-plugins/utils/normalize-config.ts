import { Common, Plugin, SDK } from '@rsdoctor/types';
import type {
  RuleSetCondition as RspackRuleSetCondition,
  RuleSetRule as RspackRuleSetRule,
} from '@rspack/core';
import {
  RuleSetCondition as WebpackRuleSetCondition,
  RuleSetConditionAbsolute as WebpackRuleSetConditionAbsolute,
  RuleSetRule as WebpackRuleSetRule,
} from 'webpack';
import {
  BriefModeConfig,
  BriefModeOptions,
  IOutput,
  NewReportCodeType,
  NormalModeOptions,
} from '@/types';

/**
 * Process mode-specific configurations with priority logic
 */
export function processModeConfigurations(
  finalMode: keyof typeof SDK.IMode,
  output: IOutput,
  brief: SDK.BriefConfig | undefined,
) {
  let finalBrief = {};
  let finalNormalOptions: NormalModeOptions = {};

  if (finalMode === 'brief') {
    finalBrief = processBriefHtmlModeConfig(output as BriefModeConfig, brief);
  } else if (finalMode === 'normal') {
    finalNormalOptions = {};
  }

  return { finalBrief, finalNormalOptions };
}

/**
 * Process brief mode configuration with priority logic
 * Priority: output.options.briefOptions > output.brief > default
 */
export function processBriefHtmlModeConfig(
  output: BriefModeConfig,
  brief: SDK.BriefConfig | undefined,
): BriefModeOptions {
  const briefOptions = output?.options as BriefModeOptions;
  const outputBriefOptions = briefOptions?.htmlOptions;
  const outputBrief = brief;

  const finalBriefOptions = outputBriefOptions ||
    outputBrief || {
      reportHtmlName: undefined,
      writeDataJson: false,
    };

  return {
    type: ['html'],
    htmlOptions: finalBriefOptions,
  };
}

/**
 * Convert reportCodeType object to NewReportCodeType enum value
 */
export function convertReportCodeTypeObject(
  reportCodeType: any,
): NewReportCodeType | undefined {
  if (!reportCodeType) return undefined;

  if (reportCodeType.noCode) {
    return 'noCode';
  } else if (reportCodeType.noAssetsAndModuleSource) {
    return 'noAssetsAndModuleSource';
  } else if (reportCodeType.noModuleSource) {
    return 'noModuleSource';
  }

  return undefined;
}

// --For loader configs--
/**
 * This function recursively processes rule set conditions to ensure they can be
 * properly serialized to JSON.
 *
 * @param item - The rule set condition to make serializable. Can be:
 *   - RspackRuleSetCondition: Rspack-specific rule conditions
 *   - WebpackRuleSetConditionAbsolute: Webpack absolute rule conditions
 *   - WebpackRuleSetCondition: Webpack rule conditions
 *   - void: Undefined or null values
 *
 * @example
 * ```typescript
 * const condition = /\.js$/;
 * JSON.stringify(condition); // Error: Converting circular structure to JSON
 *
 * makeRuleSetSerializable(condition);
 * JSON.stringify(condition); // '"/\\.js$/"'
 * ```
 */
export function makeRuleSetSerializable(
  item:
    | RspackRuleSetCondition
    | WebpackRuleSetConditionAbsolute
    | WebpackRuleSetCondition
    | void,
) {
  if (!item) return;
  if (item instanceof RegExp) {
    // Used by the JSON.stringify method to enable the transformation of an object's data for JavaScript Object Notation (JSON) serialization.
    (item as Common.PlainObject).toJSON = item.toString;
  } else if (Array.isArray(item)) {
    item.forEach((i) => makeRuleSetSerializable(i));
  } else if (typeof item === 'object') {
    makeRuleSetSerializable(item.and);
    makeRuleSetSerializable(item.or);
    makeRuleSetSerializable(item.not);
  }
}
export function makeRulesSerializable(
  rules:
    | Plugin.RuleSetRule[]
    | RspackRuleSetRule['oneOf']
    | WebpackRuleSetRule['oneOf'],
) {
  if (!Array.isArray(rules)) return;
  if (!rules.length) return;
  rules.forEach((rule) => {
    if (!rule) return;
    makeRuleSetSerializable(rule.test);
    makeRuleSetSerializable(rule.resourceQuery);
    makeRuleSetSerializable(rule.resource);
    makeRuleSetSerializable(rule.resourceFragment);
    makeRuleSetSerializable(rule.scheme);
    makeRuleSetSerializable(rule.issuer);
    if ('issuerLayer' in rule) {
      makeRuleSetSerializable(rule.issuerLayer);
    }
    makeRuleSetSerializable(rule.include);
    makeRuleSetSerializable(rule.exclude);
    if (rule.oneOf) {
      makeRulesSerializable(rule.oneOf);
    }
    if ('rules' in rule && rule.rules) {
      makeRulesSerializable(rule.rules);
    }
  });
}
