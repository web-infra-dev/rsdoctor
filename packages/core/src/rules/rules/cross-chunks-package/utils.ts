export function getErrorMsgForDupPckChunks(
  chunks: string[],
  pkgName: string,
): string {
  let message = `The same package ${pkgName} was bundled into different chunks:\n`;

  for (const chunkName of chunks) {
    message += ` ${chunkName}\n`;
  }

  return message.slice(0, -1);
}
