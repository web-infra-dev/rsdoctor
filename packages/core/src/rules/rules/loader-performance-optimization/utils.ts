export function match(str: string, patterns: (string | RegExp)[]) {
  if (patterns.length === 0) return false;

  return patterns.some((p) => {
    if (typeof p === 'string') return str === p;
    if (p instanceof RegExp) return p.test(str);
    return false;
  });
}
