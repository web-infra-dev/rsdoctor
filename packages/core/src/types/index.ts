export * from './chunks';
export * from './rules';
export * from './loader';
export * from './plugin';
