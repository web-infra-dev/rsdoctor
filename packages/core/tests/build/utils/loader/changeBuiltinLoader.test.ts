import { describe, it, expect } from 'vitest';
import os from 'os';
import { Plugin } from '@rsdoctor/types';
import { Utils } from '@/build-utils/build';
import { addProbeLoader2Rules } from '@/build-utils/build/utils';

const rules = [
  {
    test: /\.less$/,
    use: 'less-loader',
    type: 'css',
  },
  {
    test: /\.module\.less$/,
    use: 'less-loader',
    type: 'css/module',
  },
  {
    test: /\.svg$/,
    use: '@svgr/webpack',
  },
  {
    test: /\.tsx$/,
    oneOf: [
      {
        loader: 'builtin:swc-loader',
        options: {
          sourceMap: true,
          jsc: {
            parser: {
              syntax: 'typescript',
              jsx: true,
            },
          },
        },
        type: 'javascript/auto',
      },
    ],
  },
  {
    test: /\.tsx$/,
    rules: [
      {
        loader: 'builtin:swc-loader',
        options: {
          sourceMap: true,
          jsc: {
            parser: {
              syntax: 'typescript',
              jsx: true,
            },
          },
        },
        type: 'javascript/auto',
      },
    ],
  },
  {
    test: /\.ts$/,
    use: {
      loader: 'builtin:swc-loader',
      options: {
        sourceMap: true,
        jsc: {
          parser: {
            syntax: 'typescript',
          },
          externalHelpers: true,
          preserveAllComments: false,
        },
      },
    },
    type: 'javascript/auto',
  },
  {
    test: /\.svg$/,
    type: 'asset/resource',
  },
];

const mockCompiler: Plugin.BaseCompiler = {
  options: {
    name: 'test-compiler',
  },
} as Plugin.BaseCompiler;

describe('test src/build/utils/loader.ts addProbeLoader2Rules', () => {
  it('addProbeLoader2Rules()', () => {
    expect(
      addProbeLoader2Rules(rules, mockCompiler, (r: Plugin.BuildRuleSetRule) =>
        Utils.getLoaderNameMatch(r, 'builtin:swc-loader', true),
      ),
    ).toMatchSnapshot();
  });
});

describe('addProbeLoader2Rules', () => {
  const mockCompiler: Plugin.BaseCompiler = {
    options: {
      name: 'test-compiler',
    },
  } as Plugin.BaseCompiler;

  const mockRule = {
    loader: 'mock-loader',
    options: { foo: 'bar' },
  } as {
    loader: string;
    options: any;
    use?: any;
  };

  const mockFn = (rule: Plugin.BuildRuleSetRule) =>
    rule.loader === 'mock-loader';

  it('should add probe loaders to rules', () => {
    const rules = [mockRule];
    const result = addProbeLoader2Rules(rules, mockCompiler, mockFn);
    const probeLoaderPath =
      os.EOL === '\n' ? 'build-utils/build/loader/probeLoader' : 'probeLoader';

    expect(result).toHaveLength(1);
    expect(result[0]).toHaveProperty('use');
    expect(result[0].use).toHaveLength(3);
    expect(result[0].use[0]).toHaveProperty('loader');
    expect(result[0].use[0].loader).toContain(probeLoaderPath);
    console.log('result[0].use[0].loader::', result[0].use[0].loader);
    expect(result[0].use[0]).toHaveProperty('options.type', 'end');
    expect(result[0].use[1]).toHaveProperty('loader', 'mock-loader');
    expect(result[0].use[2]).toHaveProperty('loader');
    expect(result[0].use[2].loader).toContain(probeLoaderPath);
    console.log('result[0].use[2].loader::', result[0].use[2].loader);
    expect(result[0].use[2]).toHaveProperty('options.type', 'start');
  });

  it('should handle nested rules', () => {
    const nestedRule = {
      rules: [mockRule],
    };
    const rules = [nestedRule];
    const result = addProbeLoader2Rules(rules, mockCompiler, mockFn);

    expect(result).toHaveLength(1);
    expect(result[0]).toHaveProperty('rules');
    expect(result[0].rules).toHaveLength(1);
    expect(result[0].rules[0]).toHaveProperty('use');
    // @ts-ignore
    expect(result[0].rules[0].use).toHaveLength(3);
  });
});
