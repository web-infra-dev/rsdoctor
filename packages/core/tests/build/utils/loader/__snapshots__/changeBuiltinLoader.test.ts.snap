// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`test src/build/utils/loader.ts changeBuiltinLoader > changeBuiltinLoader() 1`] = `
[
  {
    "test": /\\\\\\.less\\$/,
    "type": "css",
    "use": "less-loader",
  },
  {
    "test": /\\\\\\.module\\\\\\.less\\$/,
    "type": "css/module",
    "use": "less-loader",
  },
  {
    "test": /\\\\\\.svg\\$/,
    "use": "@svgr/webpack",
  },
  {
    "oneOf": [
      {
        "loader": undefined,
        "options": undefined,
        "type": "javascript/auto",
        "use": [
          {
            "loader": "probe",
            "options": {
              "loader": "builtin:swc-loader",
              "options": {
                "jsc": {
                  "parser": {
                    "jsx": true,
                    "syntax": "typescript",
                  },
                },
                "sourceMap": true,
              },
              "type": "end",
            },
          },
          {
            "loader": "builtin:swc-loader",
            "options": {
              "jsc": {
                "parser": {
                  "jsx": true,
                  "syntax": "typescript",
                },
              },
              "sourceMap": true,
            },
          },
          {
            "loader": "probe",
            "options": {
              "loader": "builtin:swc-loader",
              "options": {
                "jsc": {
                  "parser": {
                    "jsx": true,
                    "syntax": "typescript",
                  },
                },
                "sourceMap": true,
              },
              "type": "start",
            },
          },
        ],
      },
    ],
    "test": /\\\\\\.tsx\\$/,
  },
  {
    "rules": [
      {
        "loader": undefined,
        "options": undefined,
        "type": "javascript/auto",
        "use": [
          {
            "loader": "probe",
            "options": {
              "loader": "builtin:swc-loader",
              "options": {
                "jsc": {
                  "parser": {
                    "jsx": true,
                    "syntax": "typescript",
                  },
                },
                "sourceMap": true,
              },
              "type": "end",
            },
          },
          {
            "loader": "builtin:swc-loader",
            "options": {
              "jsc": {
                "parser": {
                  "jsx": true,
                  "syntax": "typescript",
                },
              },
              "sourceMap": true,
            },
          },
          {
            "loader": "probe",
            "options": {
              "loader": "builtin:swc-loader",
              "options": {
                "jsc": {
                  "parser": {
                    "jsx": true,
                    "syntax": "typescript",
                  },
                },
                "sourceMap": true,
              },
              "type": "start",
            },
          },
        ],
      },
    ],
    "test": /\\\\\\.tsx\\$/,
  },
  {
    "test": /\\\\\\.ts\\$/,
    "type": "javascript/auto",
    "use": [
      {
        "loader": "probe",
        "options": {
          "loader": "builtin:swc-loader",
          "options": {
            "jsc": {
              "externalHelpers": true,
              "parser": {
                "syntax": "typescript",
              },
              "preserveAllComments": false,
            },
            "sourceMap": true,
          },
          "type": "end",
        },
      },
      {
        "loader": "builtin:swc-loader",
        "options": {
          "jsc": {
            "externalHelpers": true,
            "parser": {
              "syntax": "typescript",
            },
            "preserveAllComments": false,
          },
          "sourceMap": true,
        },
      },
      {
        "loader": "probe",
        "options": {
          "loader": "builtin:swc-loader",
          "options": {
            "jsc": {
              "externalHelpers": true,
              "parser": {
                "syntax": "typescript",
              },
              "preserveAllComments": false,
            },
            "sourceMap": true,
          },
          "type": "start",
        },
      },
    ],
  },
  {
    "test": /\\\\\\.svg\\$/,
    "type": "asset/resource",
  },
]
`;
